### DISCUSSION TITLE
How to add custom collector?

### DISCUSSION THREAD

**Discussion by FixRM** [PrometheusClientNet/Prometheus.Client#232] [Q&A] [answered]
Hello, I'm trying to export existing monitoring statistics in Prometheus format using this library. I tried to create custom `ICollector` implementation and add it on startup like this: `Metrics.DefaultCollectorRegistry.Add(new MyCollector());`. The problem is that I got internal errors while writing to `IMetricsWriter` as it is not clear what methods should I call and in what order. Can you please point me to documentation or existing sample? I tried to figure that out from the Gauge sources, but It seems to be too complicated for me)

p.s. May be there is a better option?

**Comment by phnx47** [comment, marked as answer]
@FixRM Hi, did you look standard collectors, example [GCCollectionCountCollector](https://github.com/prom-client-net/prom-client/blob/main/src/Prometheus.Client/Collectors/DotNetStats/GCCollectionCountCollector.cs)?

**Comment by FixRM** [comment]
Hi @phnx47! Thank you, it works! Can you please explain me the purpose of each entity in hierarchy Metric -> Sample -> Label. For instance, it is not clear what is `string value` parameter of `WriteLabel`.

**Reply by phnx47** [reply]
@FixRM WriteLabel - label name and label value. Maybe this https://prometheus.io/docs/concepts/data_model/#notation can help?

**Reply by FixRM** [reply]
Thanks. It is my first rodeo with Prometheus ( Now it is a little bit more clear. The confusing part was that there is Start/Stop labels statement but I can't write metrics value there...  Basically I'm trying to port this Prometheus.Net-based implementation to your library: https://github.com/fiftyonefifty/Hangfire.Prometheus/blob/e6e89fcd9007869199be93a3395d5ce6ea0023e2/src/Hangfire.Prometheus/Hangfire.Prometheus/PrometheusExporter.cs#L32. I think now I have everything that need, thank you.

**Reply by phnx47** [reply]
@FixRM `ICollector` is low-level api, you can use high level as well, examples: [prom-client-healthchecks](https://github.com/prom-client-net/prom-client-healthchecks) and [prom-client-httprequestdurations](https://github.com/prom-client-net/prom-client-httprequestdurations)

**Reply by FixRM** [reply]
I guess I can, but I don't really need collecting/calculating any metrics. I only need to print existing results when /metrics page is requested. Did I choose the shortest way for my purpose?

p.s. I think there was easier way than using this library at all, but at the begging I was not even familiar with Prometheus, so I used existing code to start with.

**Reply by phnx47** [reply]
Both of them are fine. With low-level `ICollector` you work with `IMetricsWriter`, with high-level you can use `Gauge` or `Untyped` metrics, like this:
```c#
var gauge = metricFactory.CreateGauge("gauge", "help text");
gauge.Set(5);
```

**Reply by FixRM** [reply]
Ok, thanks. That is entry point for high level API in my case? How can I understand that it's time to update my metrics?

**Reply by phnx47** [reply]
@FixRM It depends what you need, in your case is HangFire, check with [Hangfire.Prometheus](https://github.com/fiftyonefifty/Hangfire.Prometheus) how they update.

**Reply by FixRM** [reply]
It is library what I ported to your framework ) It use `Metrics.DefaultRegistry.AddBeforeCollectCallback` to add gauge in the way you suggested, but this method is not available in your port.

**Reply by phnx47** [reply]
I checked code in `Hangfire.Prometheus`

With `ICollector` - metrics will be collected with every request from Promethues. Similar `Hangfire.Prometheus`.
With `Gauge` - you need periodically set updated values yourself. If you have any trigger from Hangfire, you can use it or some Backgroud Job.
